[#3] 로그인

qxent300 opened the pull request:
## 변경사항
- 아이디, 비밀번호를 통한 로그인 로직 구현
- DB에서 조회하여 일치하는 데이터가 있는 경우 해당 회원의 정보를, 데이터가 없는 경우 500 에러를 리턴

qxent300 committed:
feat: 아이디, 비밀번호를 입력하여 로그인하는 로직 추가

아이디, 비밀번호를 입력받아 DB에서 조회하여 일치하는 데이터가 있는 경우 해당 회원 정보를, 데이터가 없는 경우 500 에러를 리턴.

qxent300 committed:
feat: 로그인 페이지 추가

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberController.java:
```diff
@@ -40,4 +35,20 @@ public ResponseEntity<MemberDto> joinMember(@RequestBody MemberDto memberDto) th
         return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(e.getMessage());
     }
+
+    @GetMapping("/members/login")
+    public ModelAndView login() {
+        return new ModelAndView("members/login");
+    }
```
위 controller는 view 를 보여주기 위함입니다.
현재 html 코드가 서버쪽에 의존성 없이 작성되어 있으므로, 위와같이 작성하지 않고, 파일 위치를 resources/static 으로 옮기는 것 만으로도 보여질거에요.

qxent300 replied:
네 멘토님.
위 코드는 String 타입으로 다시 수정하겠습니다.

f-lab-neo replied:
resources/static 에 파일을 넣는것 만으로도 접근이 가능합니다!

qxent300 replied:
멘토님, 말씀하신 경로로 파일 위치를 수정하게 되면 접근할때 index.html에서 a 태그에 /login.html을 설정해야 이동이 되는데, 이렇게 되면 url의 마지막에 /login.html 로 보여지게 되서 REST 규칙(?)에 맞지 않게 되는 것이 아닌가요?

f-lab-neo replied:
Restful 하다는 것은, 정적인 파일을 불러올 때 적용되는 규약이 아니라 api에서의 규약입니다. 
Html파일은 view의 영역이고 api가 아니므로, 별도로 분리하여 생각해주셔야합니다!

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberController.java:
```diff
@@ -40,4 +35,20 @@ public ResponseEntity<MemberDto> joinMember(@RequestBody MemberDto memberDto) th
+    @ExceptionHandler(value = EmptyResultDataAccessException.class)
+    public ResponseEntity<String> handleException(EmptyResultDataAccessException e) {
+        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(e.getMessage());
+    }
```
excetionHandler가 controller 내부에 있으면 적용범위는 controller에만 한정됩니다.
나중에 controller가 많이 생긴다면 어떻게 처리해야할까요?

controllerAdvice에 대하여 한번 알아보면 좋을 것 같습니다!

qxent300 replied:
네 알겠습니다.
controllerAdvice 어노테이션으로 관리하도록 수정하겠습니다.

f-lab-neo commented on src/main/java/com/project/acre/controller/MemberController.java:
```diff
@@ -40,4 +35,20 @@ public ResponseEntity<MemberDto> joinMember(@RequestBody MemberDto memberDto) th
+
+    @ExceptionHandler(value = EmptyResultDataAccessException.class)
+    public ResponseEntity<String> handleException(EmptyResultDataAccessException e) {
+        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(e.getMessage());
```
internal server error (5xx) 와 4xx 에러의 차이점은 무엇일까요?

데이터가 없는게 서버의 에러인지, 비즈니스상의 에러인지 한번 생각해보면 좋을 것 같습니다!

qxent300 replied:
제 생각에는 DB에 데이터가 없으니 서버 쪽의 문제라고 생각하였는데, 로그인 하는 클라이언트에서 잘못된 값을 입력하여서 없는 데이터를 리턴하려고 하기 때문에 400 에러가 되는것일까요?

f-lab-neo replied:
어떻게 생각하느냐에 따라 다르겠지만, 일반적으로는 동식님이 말씀하신 방향으로 처리되지 않을까 생각합니다 :)

f-lab-neo commented on src/main/java/com/project/acre/service/MemberService.java:
```diff
@@ -43,4 +40,24 @@ public MemberDto join(MemberDto memberDto) throws NoSuchAlgorithmException {
+        Optional<Member> result = memberRepository.findMember(id, sha256.encrypt(password));
+
+        if (result.isEmpty()) {
+            throw new EmptyResultDataAccessException(1);
```
repository에서 데이터가 없는 경우 (EmptyResultDataAccessException) 를 try catch로 잡아 optional.empty로 반환했는데, 이걸 또 다시 EmptyResultDataAccessException로 throw 하는것은 어색합니다.
EmptyResultDataAccessException 대신에 비즈니스적 오류 (사용자를 찾을 수 없습니다) 를 포현할 수 있는 exception을 만들어보는건 어떨까요?

qxent300 replied:
네 알겠습니다.
새로운 exception을 커스터마이징 해보도록 하겠습니다.

f-lab-neo commented on src/main/java/com/project/acre/service/MemberService.java:
```diff
@@ -43,4 +40,24 @@ public MemberDto join(MemberDto memberDto) throws NoSuchAlgorithmException {
+            memberDto.setBirth(result.get().getBirth());
+            memberDto.setEmail(result.get().getEmail());
+            memberDto.setPhone(result.get().getPhone());
+            memberDto.setClassify(result.get().getClassify());
```
setter를 사용하든 대신 생성자를 이용할 수 있을까요?

qxent300 replied:
네 멘토님.
생성자를 이용하도록 수정하겠습니다.

f-lab-neo reviewed (changes requested):
수고하셨습니다!

qxent300 committed:
refactor: 커스터마이징한 Exception 및 ExceptionHandler를 관리하는 클래스 생성

1. @ControllerAdvice를 이용하여 @ExceptionHandler를 묶어서 별도의 클래스로 관리하도록 수정
2. MemberService의 login() 메서드에서 EmptyResultDataAccessException이 아닌 커스터마이징한 Exception을 throw 하도록 CustomEmptyResultException 클래스 신규 생성 및 throw 수정
3. login() 메서드에서 MemberDto를 리턴할 때 생성자를 사용하여 초기화 하도록 수정
4. MemberController에서 로그인 페이지로 이동시 ModelAndView를 사용하지 않고 파일 경로를 리턴하도록 수정
